src/microphone-stream.ts

```diff
@@ -0,0 +1,208 @@
+    this.recorder.connect(this.context.destination);
+
+    if (stream) {
+      this.setStream(stream);
+    }
+
+    process.nextTick(() => {
+      this.emit("format", {
+        channels: 1,
+        bitDepth: 32,
+        sampleRate: this.context.sampleRate,
+        signed: true,
+        float: true,
+      });
+    });
+  }
+
+  /**
+   * Sets the MediaStream.
+   *
+   * This was separated from the constructor to enable better compatibility with Safari on iOS 11.
+   *
+   * Typically the stream is only available asynchronously, but the context must be created or
+   * resumed directly in response to a user's tap on iOS.
+   *
+   * @param {MediaStream} stream https://developer.mozilla.org/en-US/docs/Web/API/MediaStream
+   * @type {function(MediaStream): void}
+   */
+  public setStream(stream: MediaStream): void {
+    this.stream = stream;
+    this.audioInput = this.context.createMediaStreamSource(stream);
+    this.audioInput.connect(this.recorder);
+
+    /**
+     * Convert and emit the raw audio data
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/ScriptProcessorNode/onaudioprocess
+     * @param {AudioProcessingEvent} e https://developer.mozilla.org/en-US/docs/Web/API/AudioProcessingEvent
+     */
+    const recorderProcess = (e: AudioProcessingEvent) => {
+      // onaudioprocess can be called at least once after we've stopped
+      if (this.recording) {
+        this.push(
+          this.objectMode
+            ? e.inputBuffer
+            : bufferFrom(e.inputBuffer.getChannelData(0).buffer)
+        );
+      }
+    };
+    this.recorder.onaudioprocess = recorderProcess;
+  }
+
+  /**
+   * Temporarily stop emitting new data. Audio data recieved from the microphone
+   * after this will be dropped.
+   *
+   * Note: the underlying Stream interface has a .pause() API that causes new data
+   * to bebuffered rather than dropped.
+   */
+  public pauseRecording(): void {
+    this.recording = false;
+  }
+
+  /**
+   * Resume emitting new audio data after pauseRecording() was called.
+   */
+  public playRecording(): void {
+    this.recording = true;
+  }
+
+  /**
+   * Stops the recording.
+   *
+   * Note: Some versions of Firefox leave the recording icon in place after recording has stopped.
+   */
+  public stop(): void {
+    if (this.context.state === "closed") {
+      return;
+    }
+    try {
+      this.stream.getTracks()[0].stop();
+    } catch (ex) {
+      // This fails in some older versions of chrome. Nothing we can do about it.
+    }
+    this.recorder.disconnect();
+    if (this.audioInput) {
+      this.audioInput.disconnect();
+    }
+    try {
+      this.context.close(); // returns a promise;
+    } catch (ex) {
+      // this can also fail in older versions of chrome
+    }
+    this.recording = false;
+    this.push(null);
+    this.emit("close");
+  }
+
+  /**
+   * no-op, (flow-control doesn't really work on live audio)
+   */
+  public _read(/* bytes */): void {
```

Comment:
I don't think this needs to be `public`. It should only be called by the `Readable` class that this extends, so that's either `private` or `protected`.

Reply:
The `_read()` function is neither public nor private in @types/readable-stream: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2fee823196d4cc536fbc400086f1ccc582ebd128/types/readable-stream/index.d.ts#L87

TypeScript class members are public by default https://www.typescriptlang.org/docs/handbook/classes.html#public-by-default